lib/netext/httpext/error_codes.go

```diff
@@ -133,14 +144,16 @@ func errorCodeForError(err error) (errCode, string) {
 			// TODO: figure out how this happens
 			return defaultNetNonTCPErrorCode, err.Error()
 		}
-		if e.Op == "write" {
-			if sErr, ok := e.Err.(*os.SyscallError); ok {
-				switch sErr.Err {
-				case syscall.ECONNRESET:
-					return tcpResetByPeerErrorCode, tcpResetByPeerErrorCodeMsg
-				case syscall.EPIPE:
-					return tcpBrokenPipeErrorCode, tcpBrokenPipeErrorCodeMsg
-				}
+		if sErr, ok := e.Err.(*os.SyscallError); ok {
+			switch sErr.Unwrap() {
```

Comment:
I am confused - wouldn't this have been unwrapped in the call above? After all, [`net.OpError` has `Unwrap()`](https://golang.org/pkg/net/#OpError.Unwrap). Maybe that automatic `Unwrap()` above should only happen in the `default:` case of this big `switch`?

Reply:
Here we have syscall.ECONNRESET inside os.SyscallError inside net.OpError maybe inside something else. 

I think that for this chain specifically everything implements Unwrap now ... but it didn't when this code was added IIRC. I really don't want to put syscall.ECONNRESET in the top switch because:
1. that will mean the top switch will be forever long 
2. the windows fix here won't work or will need even MORE working around it >.>
3. The e.Op will get lost 

Moving the Unwrap to the default seems like a not bad idea ... should we also move the Cause then ? 

I am not really certain what wraps what and where and whether us moving things around will not break something as ... well most of the tests for this are syntethic - I created a specific wrappings of errors that than goes through this code, but as evident by the fact it hasn't caught some errors ... my matroshkas aren't what is genereated in the actual code of the application. So I do prefer to change this less ... if possible :D

Reply:
Moving Unwrap breaks the TestDNsError as it DNSError inside OpError ... maybe I should just move DNSError under OpError :thinking:

Reply:
> Moving the Unwrap to the default seems like a not bad idea ... should we also move the Cause then ?

Yes, this is what I meant, and it is going to make the change more minimal

Reply:
Unwrap in the begining is the same logic as Cause ... just slower and with a return path:
Cause basically says get me the inner error of my Cause error that doesn't implement the causer interface. 
What happens with the Unwrap above is that we Unwrap (and Cause) until we get to an error that no longer implements neither on (or just doesn't have inner error) and than we check it, if it is a concrete error - we return that and if not we go one level above and try that error 

Looking at the document [Unwrap](https://godoc.org/github.com/pkg/errors#Unwrap) was added in a later version to github.com/pkg/errors adn arguably if we upgrade the version we can drop the errors.Cause   :thinking: